presto-main/src/main/java/com/facebook/presto/type/DecimalSaturatedFloorCasts.java

```diff
@@ -0,0 +1,144 @@
+    }
+
+    @UsedByGeneratedCode
+    public static Slice shortDecimalToLongDecimal(long value, int sourcePrecision, int sourceScale, int resultPrecision, int resultScale)
+    {
+        return encodeUnscaledValue(bigintToBigintFloorSaturatedCast(BigInteger.valueOf(value), sourceScale, resultPrecision, resultScale));
+    }
+
+    @UsedByGeneratedCode
+    public static long longDecimalToShortDecimal(Slice value, int sourcePrecision, int sourceScale, int resultPrecision, int resultScale)
+    {
+        return bigintToBigintFloorSaturatedCast(decodeUnscaledValue(value), sourceScale, resultPrecision, resultScale).longValueExact();
+    }
+
+    @UsedByGeneratedCode
+    public static Slice longDecimalToLongDecimal(Slice value, int sourcePrecision, int sourceScale, int resultPrecision, int resultScale)
+    {
+        return encodeUnscaledValue(bigintToBigintFloorSaturatedCast(decodeUnscaledValue(value), sourceScale, resultPrecision, resultScale));
+    }
+
+    private static BigInteger bigintToBigintFloorSaturatedCast(BigInteger value, int sourceScale, int resultPrecision, int resultScale)
+    {
+        BigDecimal bigDecimal = new BigDecimal(value, sourceScale);
+        bigDecimal = bigDecimal.setScale(resultScale, FLOOR);
+        BigInteger unscaledValue = bigDecimal.unscaledValue();
+        BigInteger maxUnscaledValue = bigIntegerTenToNth(resultPrecision).subtract(ONE);
+        if (unscaledValue.compareTo(maxUnscaledValue) > 0) {
+            return maxUnscaledValue;
+        }
+        BigInteger minUnscaledValue = maxUnscaledValue.negate();
+        if (unscaledValue.compareTo(minUnscaledValue) < 0) {
+            return minUnscaledValue;
+        }
+        return unscaledValue;
+    }
+
+    public static final SqlScalarFunction DECIMAL_TO_BIGINT_SATURATED_FLOOR_CAST = SqlScalarFunction.builder(DecimalSaturatedFloorCasts.class)
+            .signature(Signature.builder()
+                    .kind(SCALAR)
+                    .operatorType(SATURATED_FLOOR_CAST)
+                    .literalParameters("source_precision", "source_scale")
+                    .argumentTypes("decimal(source_precision,source_scale)")
+                    .returnType("BIGINT")
+                    .build()
+            )
+            .implementation(b -> b
+                    .methods("shortDecimalToShortDecimal", "shortDecimalToLongDecimal", "longDecimalToShortDecimal", "longDecimalToLongDecimal")
+                    .withExtraParameters((context) -> {
+                        int sourceScale = Ints.checkedCast(context.getLiteral("source_scale"));
+                        return ImmutableList.of(sourceScale);
+                    })
+            ).build();
+
+    @UsedByGeneratedCode
+    public static long shortDecimalToBigint(long value, int sourceScale)
+    {
+        BigDecimal bigDecimal = new BigDecimal(BigInteger.valueOf(value), sourceScale);
+        return bigDecimal.setScale(0, FLOOR).longValueExact();
```

Comment:
Saturated floor cast is going to be use only in `DomainTranslator` to translate user-entered predicates. I don't think that good performance really matter here.